src/shogun/distributions/classical/GaussianDistributionNew.cpp

```diff
@@ -0,0 +1,297 @@
+}
+
+SGVector<float64_t> CGaussianDistributionNew::get_cov_diag() const
+{
+	return m_cov.get_row_vector(0);
+}
+
+SGMatrix<float64_t> CGaussianDistributionNew::get_cov_full() const
+{
+	return m_cov;
+}
+
+char* CGaussianDistributionNew::get_cov_type() const
+{
+	return m_cov_type;
+}
+
+SGMatrix<float64_t> CGaussianDistributionNew::sample(int32_t num_samples,
+		SGMatrix<float64_t> pre_samples) const
+{
+	REQUIRE(num_samples>0, "Number of samples (%d) must be positive\n",
+			num_samples);
+
+	/* use pre-allocated samples? */
+	SGMatrix<float64_t> samples;
+	if (pre_samples.matrix)
+	{
+		REQUIRE(pre_samples.num_rows==m_dimension, "Dimension of pre-samples"
+				" (%d) does not match dimension of Gaussian (%d)\n",
+				pre_samples.num_rows, m_dimension);
+
+		REQUIRE(pre_samples.num_cols==num_samples, "Number of pre-samples"
+				" (%d) does not desired number of samples (%d)\n",
+				pre_samples.num_cols, num_samples);
+
+		samples=pre_samples;
+	}
+	else
+	{
+		/* allocate memory and sample from std normal */
+		samples=SGMatrix<float64_t>(m_dimension, num_samples);
+		for (index_t i=0; i<m_dimension*num_samples; ++i)
+			samples.matrix[i]=sg_rand->std_normal_distrib();
+	}
+
+	/* map into desired Gaussian covariance */
+	Map<MatrixXd> eigen_samples(samples.matrix, samples.num_rows,
+			samples.num_cols);
+	Map<MatrixXd> eigen_L(m_cov.matrix, m_cov.num_rows, m_cov.num_cols);
+	eigen_samples=eigen_L*eigen_samples;
+
+	/* add mean */
+	Map<VectorXd> eigen_mean(m_mean.vector, m_mean.vlen);
+	eigen_samples.colwise()+=eigen_mean;
+
+	return samples;
+}
+
+void CGaussianDistributionNew::compute_cholesky(SGMatrix<float64_t> cov)
+{
+	Map<MatrixXd> eigen_cov(cov.matrix, cov.num_rows, cov.num_cols);
+	m_cov=SGMatrix<float64_t>(cov.num_rows, cov.num_cols);
+	Map<MatrixXd> eigen_L(m_cov.matrix, m_cov.num_rows, m_cov.num_cols);
+
+	/* compute cholesky */
+	LLT<MatrixXd> llt(eigen_cov);
+	if (llt.info()==NumericalIssue)
+	{
+		/* try to compute smallest eigenvalue for information */
+		SelfAdjointEigenSolver<MatrixXd> solver(eigen_cov);
+		if (solver.info() == Success)
+		{
+			VectorXd ev=solver.eigenvalues();
+			SG_ERROR("Error computing Cholesky of Gaussian's covariance. "
+					"Smallest Eigenvalue is %f.\n", ev[0]);
+		}
+		else
+			SG_ERROR("Error computing Cholesky of Gaussian's covariance. "
+					"Numerical Issue (could not compute eigenvalues)");
+	}
+	eigen_L = llt.matrixL();
+}
+
+SGVector<float64_t> CGaussianDistributionNew::log_pdf_multiple(SGMatrix<float64_t> samples) const
+{
+	REQUIRE(samples.num_cols>0, "Number of samples must be positive, but is %d\n",
+			samples.num_cols);
+	REQUIRE(samples.num_rows==m_dimension, "Sample dimension (%d) does not match"
+			"Gaussian dimension (%d)\n", samples.num_rows, m_dimension);
+
+	/* for easier to read code */
+	index_t num_samples=samples.num_cols;
+
+	float64_t const_part=-0.5 * m_dimension * CMath::log(2 * CMath::PI);
+
+	/* determinant is product of diagonal elements of triangular matrix */
```

Comment:
the whole method is only correct for m_cov_type == COV_FULL.

I suggest (as above) to add static methods to the class that compute things based on what the covariance is, and then add a switch over m_cov_type here and call the helper methods